File: modules/tools/whl-tools/image_message_publisher.py
#!/usr/bin/env python3
"""Publish image directories or videos as raw RGB Cyber Image messages."""

import argparse
import json
import os
import shutil
import subprocess
import sys
import time
from pathlib import Path


IMAGE_EXTENSIONS = {
    ".bmp", ".jpeg", ".jpg", ".png", ".ppm", ".tif", ".tiff", ".webp",
}


def probe_video(path):
    result = subprocess.run(
        [
            "ffprobe", "-v", "error", "-select_streams", "v:0",
            "-show_entries", "stream=width,height,r_frame_rate", "-of", "json",
            str(path),
        ],
        check=False,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        raise RuntimeError(
            f"ffprobe failed for {path}: {result.stderr.strip()}")
    streams = json.loads(result.stdout).get("streams", [])
    if not streams:
        raise RuntimeError(f"no video stream found in {path}")
    stream = streams[0]
    return int(stream["width"]), int(stream["height"]), stream.get("r_frame_rate")


def decode_image(path):
    width, height, _ = probe_video(path)
    result = subprocess.run(
        [
            "ffmpeg", "-v", "error", "-i", str(path), "-frames:v", "1",
            "-f", "rawvideo", "-pix_fmt", "rgb24", "pipe:1",
        ],
        check=False,
        capture_output=True,
    )
    expected_size = width * height * 3
    if result.returncode != 0:
        raise RuntimeError(
            f"ffmpeg failed for {path}: "
            f"{result.stderr.decode(errors='replace').strip()}")
    if len(result.stdout) != expected_size:
        raise RuntimeError(
            f"decoded image size mismatch for {path}: "
            f"{len(result.stdout)} != {expected_size}")
    return width, height, result.stdout


def image_paths(directory, recursive, max_frames):
    iterator = (
        (Path(root) / name for root, _, names in os.walk(directory)
         for name in names)
        if recursive
        else (Path(entry.path) for entry in os.scandir(directory)
              if entry.is_file())
    )
    paths = []
    for path in iterator:
        if path.suffix.lower() not in IMAGE_EXTENSIONS:
            continue
        paths.append(path)
        if max_frames and len(paths) >= max_frames:
            break
    return paths


def relative_path(path, dataset_root):
    if dataset_root is None:
        return path.name
    try:
        return str(path.resolve().relative_to(dataset_root.resolve()))
    except ValueError as error:
        raise RuntimeError(
            f"{path} is outside dataset root {dataset_root}") from error


def publish_message(writer, image_type, width, height, rgb, sequence,
                    timestamp_ns, frame_id):
    message = image_type()
    message.header.timestamp_sec = timestamp_ns / 1_000_000_000
    message.header.camera_timestamp = timestamp_ns
    message.header.module_name = "whl_image_message_publisher"
    message.header.sequence_num = sequence
    message.header.frame_id = frame_id
    message.frame_id = frame_id
    message.measurement_time = timestamp_ns / 1_000_000_000
    message.height = height
    message.width = width
    message.encoding = "rgb8"
    message.step = width * 3
    message.data = rgb
    writer.write(message)


def publish_images(args, writer, image_type, manifest):
    paths = image_paths(Path(args.input), args.recursive, args.max_frames)
    if not paths:
        raise RuntimeError(f"no supported images found in {args.input}")
    interval_ns = round(1_000_000_000 / args.fps)
    for index, path in enumerate(paths, start=1):
        width, height, rgb = decode_image(path)
        timestamp_ns = args.start_timestamp_ns + (index - 1) * interval_ns
        record = {
            "sequence_num": index,
            "camera_timestamp_ns": timestamp_ns,
            "source_path": str(path.resolve()),
            "relative_path": relative_path(path, args.dataset_root),
            "source_frame_index": index - 1,
            "width": width,
            "height": height,
        }
        manifest.write(json.dumps(record, sort_keys=True) + "\n")
        manifest.flush()
        publish_message(
            writer, image_type, width, height, rgb, index, timestamp_ns,
            args.frame_id)
        print(f"published {index}: {path} at {timestamp_ns} ns", flush=True)
        time.sleep(1.0 / args.fps)


def publish_video(args, writer, image_type, manifest):
    path = Path(args.input)
    width, height, source_rate = probe_video(path)
    fps = args.fps
    if fps is None:
        numerator, denominator = source_rate.split("/", maxsplit=1)
        fps = float(numerator) / float(denominator)
    frame_size = width * height * 3
    process = subprocess.Popen(
        [
            "ffmpeg", "-v", "error", "-i", str(path), "-f", "rawvideo",
            "-pix_fmt", "rgb24", "pipe:1",
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
    )
    interval_ns = round(1_000_000_000 / fps)
    sequence = 1
    try:
        while not args.max_frames or sequence <= args.max_frames:
            rgb = process.stdout.read(frame_size)
            if not rgb:
                break
            if len(rgb) != frame_size:
                raise RuntimeError(
                    f"partial decoded video frame: {len(rgb)} != {frame_size}")
            timestamp_ns = (
                args.start_timestamp_ns + (sequence - 1) * interval_ns)
            virtual_path = f"{path.stem}/frame_{sequence:06d}.jpg"
            record = {
                "sequence_num": sequence,
                "camera_timestamp_ns": timestamp_ns,
                "source_path": str(path.resolve()),
                "relative_path": virtual_path,
                "source_frame_index": sequence - 1,
                "width": width,
                "height": height,
            }
            manifest.write(json.dumps(record, sort_keys=True) + "\n")
            manifest.flush()
            publish_message(
                writer, image_type, width, height, rgb, sequence, timestamp_ns,
                args.frame_id)
            print(
                f"published video frame {sequence} at {timestamp_ns} ns",
                flush=True)
            sequence += 1
            time.sleep(1.0 / fps)
    finally:
        process.terminate()
        process.wait()
    if sequence == 1:
        error = process.stderr.read().decode(errors="replace").strip()
        raise RuntimeError(f"no video frames decoded from {path}: {error}")


def parse_args():
    parser = argparse.ArgumentParser(
        description=(
            "Publish an image directory or video as rgb8 Cyber Image messages "
            "and write a source-ID manifest for lane evaluation."))
    parser.add_argument("--input", required=True)
    parser.add_argument(
        "--channel", default="/apollo/sensor/camera/front_6mm/image")
    parser.add_argument("--frame-id", default="camera_front_6mm")
    parser.add_argument("--fps", type=float)
    parser.add_argument("--max-frames", type=int, default=0)
    parser.add_argument("--startup-wait-sec", type=float, default=3.0)
    parser.add_argument("--start-timestamp-ns", type=int, default=0)
    parser.add_argument("--manifest", default="image_manifest.jsonl")
    parser.add_argument("--dataset-root", type=Path)
    parser.add_argument("--recursive", action="store_true")
    args = parser.parse_args()
    input_path = Path(args.input)
    if not input_path.exists():
        parser.error(f"input does not exist: {input_path}")
    if args.fps is None and input_path.is_dir():
        args.fps = 10.0
    if args.fps is not None and args.fps <= 0:
        parser.error("--fps must be positive")
    if args.max_frames < 0 or args.startup_wait_sec < 0:
        parser.error("--max-frames and --startup-wait-sec must be non-negative")
    if args.start_timestamp_ns < 0:
        parser.error("--start-timestamp-ns must be non-negative")
    if args.start_timestamp_ns == 0:
        args.start_timestamp_ns = time.time_ns()
    return args


def main():
    args = parse_args()
    if shutil.which("ffmpeg") is None or shutil.which("ffprobe") is None:
        print("ffmpeg and ffprobe are required", file=sys.stderr)
        return 2
    try:
        from cyber.python.cyber_py3 import cyber
        from wheelos_msgs.sensor_msgs import sensor_image_pb2
    except ImportError as error:
        print(f"Cyber runtime imports failed: {error}", file=sys.stderr)
        return 2

    manifest_path = Path(args.manifest)
    manifest_path.parent.mkdir(parents=True, exist_ok=True)
    try:
        with manifest_path.open("x", encoding="utf-8") as manifest:
            cyber.init("whl_image_message_publisher")
            node = cyber.Node("whl_image_message_publisher")
            writer = node.create_writer(
                args.channel, sensor_image_pb2.Image, 8)
            time.sleep(args.startup_wait_sec)
            if Path(args.input).is_dir():
                publish_images(args, writer, sensor_image_pb2.Image, manifest)
            else:
                publish_video(args, writer, sensor_image_pb2.Image, manifest)
    except (OSError, RuntimeError, ValueError) as error:
        print(error, file=sys.stderr)
        return 1
    finally:
        if "cyber" in locals():
            cyber.shutdown()
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: modules/tools/whl-tools/image_message_publisher.cc
// Copyright 2026 WheelOS. All Rights Reserved.
//
// Publish decoded images as apollo.drivers.Image messages for offline debugging.

#include <algorithm>
#include <cctype>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <filesystem>
#include <iostream>
#include <limits>
#include <string>
#include <thread>
#include <utility>
#include <vector>

#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/videoio.hpp>

#include "cyber/cyber.h"
#include "cyber/time/time.h"
#include "wheelos_msgs/sensor_msgs/sensor_image.pb.h"

namespace {

struct Options {
  std::filesystem::path input_path;
  std::string channel = "/apollo/sensor/camera/front_6mm/image";
  std::string frame_id = "camera_front_6mm";
  double fps = 10.0;
  uint64_t start_timestamp_ns = 0;
  uint32_t max_frames = 0;
};

void PrintUsage(const char* program) {
  std::cout
      << "Usage: " << program << " --input <directory|video> [options]\n"
      << "Publish images as packed rgb8 apollo.drivers.Image messages.\n\n"
      << "Options:\n"
      << "  --channel <topic>             Output channel.\n"
      << "  --frame-id <id>               Image and header frame ID.\n"
      << "  --fps <positive number>       Publish rate; video defaults to source FPS"
         " when omitted.\n"
      << "  --max-frames <count>          Stop after this many frames; 0 means all.\n"
      << "  --start-timestamp-ns <uint64> First frame sensor timestamp; defaults to"
         " current Cyber time.\n"
      << "  -h, --help                    Show this help.\n";
}

bool ParseUint64(const std::string& text, uint64_t* value) {
  if (value == nullptr || text.empty()) {
    return false;
  }
  try {
    size_t consumed = 0;
    const auto parsed = std::stoull(text, &consumed);
    if (consumed != text.size()) {
      return false;
    }
    *value = parsed;
    return true;
  } catch (const std::exception&) {
    return false;
  }
}

bool ParseUint32(const std::string& text, uint32_t* value) {
  uint64_t parsed = 0;
  if (!ParseUint64(text, &parsed) ||
      parsed > std::numeric_limits<uint32_t>::max()) {
    return false;
  }
  *value = static_cast<uint32_t>(parsed);
  return true;
}

bool ParseDouble(const std::string& text, double* value) {
  if (value == nullptr || text.empty()) {
    return false;
  }
  try {
    size_t consumed = 0;
    const double parsed = std::stod(text, &consumed);
    if (consumed != text.size() || !std::isfinite(parsed) || parsed <= 0.0) {
      return false;
    }
    *value = parsed;
    return true;
  } catch (const std::exception&) {
    return false;
  }
}

bool ParseOptions(int argc, char* argv[], Options* options) {
  if (options == nullptr) {
    return false;
  }
  bool fps_specified = false;
  for (int index = 1; index < argc; ++index) {
    const std::string argument = argv[index];
    if (argument == "-h" || argument == "--help") {
      PrintUsage(argv[0]);
      return false;
    }
    if (index + 1 == argc) {
      std::cerr << "Missing value for " << argument << "\n";
      return false;
    }
    const std::string value = argv[++index];
    if (argument == "--input") {
      options->input_path = value;
    } else if (argument == "--channel") {
      options->channel = value;
    } else if (argument == "--frame-id") {
      options->frame_id = value;
    } else if (argument == "--fps") {
      fps_specified = ParseDouble(value, &options->fps);
      if (!fps_specified) {
        std::cerr << "Invalid --fps value: " << value << "\n";
        return false;
      }
    } else if (argument == "--max-frames") {
      if (!ParseUint32(value, &options->max_frames)) {
        std::cerr << "Invalid --max-frames value: " << value << "\n";
        return false;
      }
    } else if (argument == "--start-timestamp-ns") {
      if (!ParseUint64(value, &options->start_timestamp_ns)) {
        std::cerr << "Invalid --start-timestamp-ns value: " << value << "\n";
        return false;
      }
    } else {
      std::cerr << "Unknown option: " << argument << "\n";
      return false;
    }
  }
  if (options->input_path.empty() || options->channel.empty() ||
      options->frame_id.empty()) {
    std::cerr << "--input, --channel, and --frame-id must not be empty\n";
    return false;
  }
  if (!std::filesystem::exists(options->input_path)) {
    std::cerr << "Input does not exist: " << options->input_path << "\n";
    return false;
  }
  if (std::filesystem::is_regular_file(options->input_path) && !fps_specified) {
    cv::VideoCapture video(options->input_path.string());
    const double source_fps = video.get(cv::CAP_PROP_FPS);
    if (source_fps > 0.0 && std::isfinite(source_fps)) {
      options->fps = source_fps;
    }
  }
  return true;
}

bool IsImageFile(const std::filesystem::path& path) {
  std::string extension = path.extension().string();
  std::transform(extension.begin(), extension.end(), extension.begin(),
                 [](unsigned char character) { return std::tolower(character); });
  return extension == ".bmp" || extension == ".jpeg" || extension == ".jpg" ||
         extension == ".png" || extension == ".ppm" || extension == ".tif" ||
         extension == ".tiff";
}

std::vector<std::filesystem::path> ImagePaths(
    const std::filesystem::path& directory) {
  std::vector<std::filesystem::path> paths;
  for (const auto& entry : std::filesystem::directory_iterator(directory)) {
    if (entry.is_regular_file() && IsImageFile(entry.path())) {
      paths.push_back(entry.path());
    }
  }
  std::sort(paths.begin(), paths.end());
  return paths;
}

bool PublishFrame(const cv::Mat& input, uint64_t timestamp_ns, uint32_t sequence,
                  const Options& options,
                  const std::shared_ptr<apollo::cyber::Writer<
                      apollo::drivers::Image>>& writer) {
  if (input.empty() || writer == nullptr) {
    return false;
  }
  cv::Mat rgb;
  if (input.channels() == 1) {
    cv::cvtColor(input, rgb, cv::COLOR_GRAY2RGB);
  } else if (input.channels() == 3) {
    cv::cvtColor(input, rgb, cv::COLOR_BGR2RGB);
  } else if (input.channels() == 4) {
    cv::cvtColor(input, rgb, cv::COLOR_BGRA2RGB);
  } else {
    std::cerr << "Unsupported image channel count: " << input.channels() << "\n";
    return false;
  }
  if (!rgb.isContinuous()) {
    rgb = rgb.clone();
  }
  const uint64_t byte_count =
      static_cast<uint64_t>(rgb.rows) * static_cast<uint64_t>(rgb.cols) * 3U;
  if (byte_count > std::numeric_limits<int>::max()) {
    std::cerr << "Image is too large to serialize\n";
    return false;
  }
  auto image = std::make_shared<apollo::drivers::Image>();
  const double timestamp_sec = static_cast<double>(timestamp_ns) * 1.0e-9;
  auto* header = image->mutable_header();
  header->set_timestamp_sec(timestamp_sec);
  header->set_camera_timestamp(timestamp_ns);
  header->set_module_name("whl_image_message_publisher");
  header->set_sequence_num(sequence);
  header->set_frame_id(options.frame_id);
  image->set_frame_id(options.frame_id);
  image->set_measurement_time(timestamp_sec);
  image->set_height(static_cast<uint32_t>(rgb.rows));
  image->set_width(static_cast<uint32_t>(rgb.cols));
  image->set_encoding("rgb8");
  image->set_step(static_cast<uint32_t>(rgb.cols * 3));
  image->set_data(reinterpret_cast<const char*>(rgb.data),
                  static_cast<int>(byte_count));
  writer->Write(image);
  return true;
}

template <typename NextFrame>
int PublishFrames(const Options& options, NextFrame next_frame) {
  const uint64_t interval_ns =
      static_cast<uint64_t>(std::llround(1.0e9 / options.fps));
  const uint64_t start_timestamp_ns =
      options.start_timestamp_ns == 0
          ? static_cast<uint64_t>(apollo::cyber::Time::Now().ToNanosecond())
          : options.start_timestamp_ns;
  auto node = apollo::cyber::CreateNode("whl_image_message_publisher");
  auto writer = node->CreateWriter<apollo::drivers::Image>(options.channel);
  if (writer == nullptr) {
    std::cerr << "Unable to create writer for " << options.channel << "\n";
    return 1;
  }

  uint32_t sequence = 1;
  cv::Mat frame;
  while ((options.max_frames == 0U || sequence <= options.max_frames) &&
         next_frame(&frame)) {
    if (!PublishFrame(frame, start_timestamp_ns +
                                 static_cast<uint64_t>(sequence - 1) * interval_ns,
                      sequence, options, writer)) {
      return 1;
    }
    std::cout << "Published frame " << sequence << " at "
              << start_timestamp_ns +
                     static_cast<uint64_t>(sequence - 1) * interval_ns
              << " ns\n";
    ++sequence;
    std::this_thread::sleep_for(
        std::chrono::nanoseconds(static_cast<int64_t>(interval_ns)));
  }
  if (sequence == 1) {
    std::cerr << "No decodable frames found in " << options.input_path << "\n";
    return 1;
  }
  return 0;
}

}  // namespace

int main(int argc, char* argv[]) {
  Options options;
  if (!ParseOptions(argc, argv, &options)) {
    return argc > 1 && (std::string(argv[1]) == "--help" ||
                        std::string(argv[1]) == "-h")
               ? 0
               : 2;
  }
  apollo::cyber::Init("whl_image_message_publisher");
  int result = 0;
  if (std::filesystem::is_directory(options.input_path)) {
    const auto paths = ImagePaths(options.input_path);
    size_t index = 0;
    result = PublishFrames(options, [&paths, &index](cv::Mat* frame) {
      if (index == paths.size()) {
        return false;
      }
      *frame = cv::imread(paths[index++].string(), cv::IMREAD_UNCHANGED);
      return true;
    });
  } else {
    cv::VideoCapture video(options.input_path.string());
    if (!video.isOpened()) {
      std::cerr << "Unable to open video: " << options.input_path << "\n";
      result = 1;
    } else {
      result = PublishFrames(
          options, [&video](cv::Mat* frame) { return video.read(*frame); });
    }
  }
  apollo::cyber::Clear();
  return result;
}
